### PR TITLE
Cext 4219

### PR DESCRIPTION
**Pull request by AjazSumaiya**
## Description
This PR closes - https://jira.corp.adobe.com/browse/CEXT-4219

## Motivation and Context

Current log-get-bulk command accepts startTime and endTime in UTC format to fetch logs for 30 mins time window
`aio api-mesh log-get-bulk -- startTime YYYY-MM-DDTHH: MM: SSZ -- endTime YYYY-MM-DDTHH: MM: SSZ -- filename mesh_logs. csv`

It is bit inconvenient for the end user to give time in this particular format YYYY-MM-DDTHH: MM:SSZ.

Rather user should be able to give -- past 30 instead of – startTime YYYY-MM-DDTHH: MM:SSZ – endTime YYYY-MM-
DDTHH: MM:SSZ flag, to fetch the logs without needing to convert the time in the desired format.

Two cases need to be considered with the adding of the – past flag -

1. User requests log in for any time period less then 30 mins from now

- Calculate the startTime based on the current time which will serve as endTime, and use exisiting logic to fetch the logs.
- aio api-mesh log-get-bulk -- past 30mins -- filename logs.csv

2. User requests logs for any time period but less than 30 mins in any arbitrary time frame in the past the 30 days.

- Introduced `--from flag` - the time from which the `–past` time window needs to be calculated

- aio api-mesh log-get-bulk -- past 30mins -- from YYYY-MM-DD:HH:MM:SS -- filename logs.csv
- Convert the local time to UTC and then use existing logic to fetch logs, the decided format for –from time
 is YYYY-MM-DD:HH:MM:SS.

## How Has This Been Tested?
Addition of the above flags are additive and all the existing logic with --startTime and --endTime should work as is 
- Ran unit tests locally all passed

Tested scenarios for `--past` flag - 
1. if no logs found in provided --past window
image

2. should download logs if found in given window 
image
TimeStamps should match from `aio api-mesh log-list`
image

3. Invalid `--past` duration format (e.g., `20hour`, `20s`)
image

4. Accepted valid `--past` durations abbreviations
image
image
  

Tested scenarios for `--from` flag - 
1. should download logs if found in given time window
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/commands/api-mesh/__tests__/log-get-bulk.test.js [hunk 1]
```diff
@@ -291,3 +295,249 @@ describe('GetBulkLogCommand startTime and endTime validation', () => {
...
+			expect.any(String),
+			expect.any(String),
+		);
+	});
+
+	test('throws an error with edge case for --past duration', async () => {
+		parseSpy.mockResolvedValueOnce({
+			flags: {
+				past: '0s',
+				from: '2025-03-10:12:00:00',
```
Comment (dthampy) [inline src/commands/api-mesh/__tests__/log-get-bulk.test.js:432]: same comment as above regarding the test failure once we go past 30 days from this date?
Check all places where the from is hard coded and expects to not throw an error

File: src/commands/api-mesh/__tests__/log-get-bulk.test.js [hunk 2]
```diff
@@ -291,3 +295,249 @@ describe('GetBulkLogCommand startTime and endTime validation', () => {
...
 });
+
+describe('GetBulkLogCommand with --past and --from flags', () => {
+	let parseSpy;
+
+	beforeEach(() => {
+		parseSpy = jest.spyOn(GetBulkLogCommand.prototype, 'parse').mockResolvedValue({
+			flags: {
+				past: '20mins',
+				from: '2025-03-09:12:00:00',
```
Comment (dthampy) [inline src/commands/api-mesh/__tests__/log-get-bulk.test.js:306]: Would this test fail if we go past 30 days from this date because of that 30day restriction we impose?

File: src/utils.js [hunk 3]
```diff
@@ -79,12 +80,19 @@ const fileNameFlag = Flags.string({
...
-	required: true,
+});
+
+const pastFlag = Flags.string({
+	description: 'Past time window in mins',
+});
+
+const fromFlag = Flags.string({
+	description:
+		'From time in DD-MM-YYYY:HH:MM:SS format to fetch logs from the past. Used as the starting time for the past time duration.',
```
Comment (dani97) [inline src/utils.js:95]: @AjazSumaiya can we add a note here that, from flag takes system time?

Comment (dani97) [reply src/utils.js:95]: summary: 'brief summary',                    // help summary for flag
You can add extra details here

File: src/utils.js [hunk 4]
```diff
@@ -605,6 +613,141 @@ function suggestCorrectedDateFormat(inputDate) {
...
+	const dateTimeRegex = /^(?:(\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])T(0[0-9]|1[0-9]|2[0-3]):([0-5]\d):([0-5]\d)Z)$/;
+
+	// Convert the Date object to ISO string and remove milliseconds
+	let timeString = time.toISOString().replace(/\.\d{3}Z$/, 'Z');
+
+	// Validate the formatted time string against the regex
+	if (!dateTimeRegex.test(timeString)) {
+		const correctedDate = suggestCorrectedDateFormat(timeString);
+		if (!correctedDate) {
+			throw Error(`Found invalid date components for ${timeString}. Check and correct the date.`);
```
Comment (jhadobe) [inline src/utils.js:713]:
```suggestion
			throw Error(`Invalid date components in ${timeString}. Confirm the date information is correct.`);
```

i would use something like this. Please repeat for similar error messages.

File: src/utils.js [hunk 5]
```diff
@@ -605,6 +613,141 @@ function suggestCorrectedDateFormat(inputDate) {
...
+	// Check if the duration between start and end times is greater than 30 minutes (1800 seconds)
+	const timeDifferenceInSeconds = (end.getTime() - start.getTime()) / 1000;
+
+	if (timeDifferenceInSeconds > 1800) {
+		const hours = Math.floor(timeDifferenceInSeconds / 3600); // Hours calculation
+		const minutes = Math.floor((timeDifferenceInSeconds % 3600) / 60); // Minutes calculation
+		const seconds = timeDifferenceInSeconds % 60; // Seconds calculation
+
+		throw new Error(
+			`Max duration between startTime and endTime should be 30 minutes. Current duration is ${hours} hour${
```
Comment (jhadobe) [inline src/utils.js:688]:
```suggestion
			`The maximum duration between startTime and endTime is 30 minutes. The current duration is ${hours} hour${
```

File: src/utils.js [hunk 6]
```diff
@@ -605,6 +613,141 @@ function suggestCorrectedDateFormat(inputDate) {
...
+		throw new Error('endTime must be greater than startTime');
+	}
+
+	// Validate that endTime is not greater than the current time (now)
+	if (end > now) {
+		throw new Error('endTime cannot be in the future. Provide a valid endTime.');
+	}
+
+	if (start.getTime() === end.getTime()) {
+		throw new Error('Minimum duration should be 1 minutes. Current duration is 0 minutes.');
```
Comment (jhadobe) [inline src/utils.js:676]:
```suggestion
		throw new Error('The minimum duration is 1 minutes. The current duration is 0 minutes.');
```

File: src/utils.js [hunk 7]
```diff
@@ -605,6 +613,141 @@ function suggestCorrectedDateFormat(inputDate) {
...
+ * @returns {number} The duration in milliseconds.
+ */
+function parsePastDuration(pastTimeWindow) {
+	// Regular expression to match various formats of minute abbreviations
+	const pastDurationRegex = /^(\d+)\s*(m|mins?|minutes?)$/i;
+	const match = pastTimeWindow.match(pastDurationRegex);
+
+	if (!match) {
+		throw new Error(
+			'Invalid format. Past time window should be in minutes, e.g., "20 mins", "15 minutes".',
```
Comment (jhadobe) [inline src/utils.js:629]:
```suggestion
			'Invalid format. The past time window should be in minutes, for example, "20 mins", "15 minutes".',
```

File: src/utils.js (hunk 3 above)
Comment (jhadobe) [inline src/utils.js:95]:
```suggestion
		'The from time in DD-MM-YYYY:HH:MM:SS format. It is used to fetch logs from the past and is the starting time for the past time duration.',
```

File: src/commands/api-mesh/log-get-bulk.js [hunk 8]
```diff
@@ -35,64 +43,88 @@ class GetBulkLogCommand extends Command {
...
+				let convertedTime;
+				const dateTimeRegex = /^\d{4}-\d{2}-\d{2}:\d{2}:\d{2}:\d{2}$/;
+				if (!dateTimeRegex.test(flags.from)) {
+					this.error('Invalid format. Use the format YYYY-MM-DD:HH:MM:SS for --from.');
+				} else {
+					try {
+						convertedTime = await localToUTCTime(flags.from.toString());
+					} catch (error) {
+						this.error(
+							`Found invalid date components passed in --from. Check and correct the date.`,
```
Comment (jhadobe) [inline src/commands/api-mesh/log-get-bulk.js:103]:
```suggestion
							`Invalid date components passed in --from. Correct the date.`,
```

File: src/commands/api-mesh/__tests__/log-get-bulk.test.js [hunk 9]
```diff
@@ -291,3 +295,267 @@ describe('GetBulkLogCommand startTime and endTime validation', () => {
...
+		},
+		{
+			startTime: new Date(new Date().setUTCDate(new Date().getUTCDate() - 31)).toISOString(),
+			endTime: '2025-03-10T12:00:00Z',
+			error: 'Cannot get logs more than 30 days old. Adjust your time range.',
+		},
+		{
+			startTime: '2025-03-09T12:00:00Z',
+			endTime: '2025-03-09T12:00:00Z',
+			error: 'Minimum duration should be 1 minutes. Current duration is 0 minutes.',
```
Comment (jhadobe) [inline src/commands/api-mesh/__tests__/log-get-bulk.test.js:503]:
```suggestion
			error: 'The minimum duration is minutes. Current duration is 0 minutes.',
```

File: src/commands/api-mesh/__tests__/log-get-bulk.test.js [hunk 10]
```diff
@@ -291,3 +295,267 @@ describe('GetBulkLogCommand startTime and endTime validation', () => {
...
+				past: '20mins',
+				from: fromDate.toISOString().slice(0, 10) + ':25:61:61',
+				filename: 'test.csv',
+				ignoreCache: false,
+			},
+		});
+
+		const command = new GetBulkLogCommand([], {});
+		await expect(command.run()).rejects.toThrow(
+			'Found invalid date components passed in --from. Check and correct the date.',
```
Comment (jhadobe) [inline src/commands/api-mesh/__tests__/log-get-bulk.test.js:388]:
```suggestion
			'Invalid date components passed in --from. Correct the date.',
```

**Review by jhadobe — CHANGES_REQUESTED** [review]

File: src/utils.js (hunk 3 above)
Comment (dani97) [reply src/utils.js:95]: @jhadobe can you suggest how we can inform the customer that this time input  would be inferred based on system timezone.

File: src/utils.js (hunk 4 above)
Comment (AjazSumaiya) [reply src/utils.js:713]: sure @jhadobe, will update all the messages

File: src/utils.js (hunk 3 above)
Comment (jhadobe) [reply src/utils.js:95]:
```suggestion
		'The from time in DD-MM-YYYY:HH:MM:SS format based on your system's time zone. It is used to fetch logs from the past and is the starting time for the past time duration.',
```
@dani97 i would suggest this if you mean the user's system timezone.

**Review by jhadobe — APPROVED** [review]

File: src/utils.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (AjazSumaiya) [reply src/utils.js]: yeah it should be YYYY-MM-DD:HH:MM:SS , i'll change

**Review by kmaschi — APPROVED** [review]
